pkg/network/protocols/http/protocol.go

```diff
@@ -0,0 +1,215 @@
+		statkeeper: statkeeper,
+	}, nil
+}
+
+func (p *httpProtocol) ConfigureOptions(mgr *manager.Manager, opts *manager.Options) {
+	if p == nil {
+		return
+	}
+
+	opts.MapSpecEditors[httpInFlightMap] = manager.MapSpecEditor{
+		Type:       ebpf.Hash,
+		MaxEntries: uint32(p.cfg.MaxTrackedConnections),
+		EditorFlag: manager.EditMaxEntries,
+	}
+
+	// Configure event stream
+	events.Configure("http", mgr, opts)
+}
+
+func (p *httpProtocol) PreStart(mgr *manager.Manager) (err error) {
+	if p == nil {
+		return
+	}
+
+	p.eventsConsumer, err = events.NewConsumer(
+		"http",
+		mgr,
+		p.processHTTP,
+	)
+	if err != nil {
+		return
+	}
+
+	p.eventsConsumer.Start()
+
+	return
+}
+
+func (p *httpProtocol) PostStart(mgr *manager.Manager) {
+	if p == nil {
+		return
+	}
+
+	p.setupMapCleaner(mgr)
+
+	log.Info("http monitoring enabled")
+}
+
+func (p *httpProtocol) PreStop(mgr *manager.Manager) {
+	if p == nil {
+		return
+	}
+
+	p.mapCleaner.Stop()
+	p.eventsConsumer.Stop()
+	p.statkeeper.Close()
+}
+
+func (p *httpProtocol) PostStop(mgr *manager.Manager) {
+}
+
+func (p *httpProtocol) DumpMaps(output *strings.Builder, mapName string, currentMap *ebpf.Map) {
+	if p == nil {
+		return
+	}
+
+	switch mapName {
+	case httpInFlightMap: // maps/http_in_flight (BPF_MAP_TYPE_HASH), key ConnTuple, value httpTX
+		output.WriteString("Map: '" + mapName + "', key: 'ConnTuple', value: 'httpTX'\n")
+		iter := currentMap.Iterate()
+		var key netebpf.ConnTuple
+		var value HttpTX
+		for iter.Next(unsafe.Pointer(&key), unsafe.Pointer(&value)) {
+			output.WriteString(spew.Sdump(key, value))
+		}
+
+	}
+}
+
+func (p *httpProtocol) processHTTP(data []byte) {
+	if p == nil {
+		return
+	}
+
+	tx := (*EbpfHttpTx)(unsafe.Pointer(&data[0]))
+	p.telemetry.Count(tx)
+	p.statkeeper.Process(tx)
+}
+
+func (p *httpProtocol) setupMapCleaner(mgr *manager.Manager) {
+	if p == nil {
+		return
+	}
+
+	httpMap, _, _ := mgr.GetMap(httpInFlightMap)
+	mapCleaner, err := ddebpf.NewMapCleaner(httpMap, new(netebpf.ConnTuple), new(EbpfHttpTx))
+	if err != nil {
+		log.Errorf("error creating map cleaner: %s", err)
+		return
+	}
+
+	ttl := p.cfg.HTTPIdleConnectionTTL.Nanoseconds()
+	mapCleaner.Clean(p.cfg.HTTPMapCleanerInterval, func(now int64, key, val interface{}) bool {
+		httpTxn, ok := val.(*EbpfHttpTx)
+		if !ok {
+			return false
+		}
+
+		if updated := int64(httpTxn.ResponseLastSeen()); updated > 0 {
+			return (now - updated) > ttl
+		}
+
+		started := int64(httpTxn.RequestStarted())
+		return started > 0 && (now-started) > ttl
+	})
+
+	p.mapCleaner = mapCleaner
+}
+
+// GetStats returns a map of HTTP stats stored in the following format:
+// [source, dest tuple, request path] -> RequestStats object
+func (p *httpProtocol) GetStats() *protocols.ProtocolStats {
+	if p == nil {
+		return nil
+	}
+
+	p.eventsConsumer.Sync()
+	p.telemetry.Log()
+	return &protocols.ProtocolStats{
+		Type:  protocols.HTTP,
+		Stats: p.statkeeper.GetAndResetAllStats(),
+	}
+}
+
+func init() {
```

Comment:
I agree it is not ideal. I wrote it when I started this refactoring, and the new protocol interfaces where in the USM module (meaning it could create a circular dependency). This is no longer an issue of course, but it worked well enough and, I was curious to get the input of people on it :)

Looking at it now, I think I can try doing that without a global map, nor use of the `init` function, which I agree would be cleaner.

Reply:
Moved to the monitor package now, and the use of init has been removed.